dashboard/src/modules/components/PaginationComponent/index.jsx

```diff
@@ -1,28 +1,79 @@
+import * as TYPES from "actions/types";
+
 import { Pagination, PaginationVariant } from "@patternfly/react-core";
+import {
+  fetchPublicDatasets,
+  setPageLimit,
+  setPerPage,
+} from "actions/datasetListActions";
+import { useDispatch, useSelector } from "react-redux";
+
+import { LIMIT_MULTIPLIER } from "assets/constants/browsingPageConstants";
 import React from "react";
 
-const TablePagination = ({
-  numberOfRows,
-  page,
-  setPage,
-  perPage,
-  setPerPage,
-}) => {
+const TablePagination = ({ page, setPage }) => {
+  const dispatch = useDispatch();
+
+  const { totalDatasets, publicData, perPage } = useSelector(
+    (state) => state.datasetlist
+  );
   const onSetPage = (_event, pageNumber) => {
     setPage(pageNumber);
+
+    fetchData(_event, pageNumber, perPage);
+  };
+  const onPerPageSelect = (_event, newPerPage, newPage) => {
+    dispatch(setPageLimit(newPerPage * LIMIT_MULTIPLIER));
+
+    dispatch(setPerPage(newPerPage));
+    setPage(newPage);
+
+    fetchData(_event, newPage, newPerPage);
   };
-  const onPerPageSelect = (_event, perPage) => {
-    setPerPage(perPage);
+  const fetchData = (_event, newPage, newPerPage = perPage) => {
+    const startIdx = (newPage - 1) * newPerPage;
+    const endIdx = newPage * newPerPage;
+    let left = startIdx;
+    let right = endIdx;
+    while (left < right) {
+      if (publicData[startIdx]) {
+        left++;
+      } else {
+        break;
+      }
+      if (publicData[endIdx]) {
+        right--;
+      } else {
+        break;
+      }
+    }
+    if (left !== right) {
```

Comment:
I'm not sure I understand what's happening here. We're searching from both ends looking for data within the slice of the sparse `publicData` array representing the current page. But ... why? I understand that the page may not be full (though that should only happen if we're on the *last* page, right?), but if we've fetched data for the page already then `publicData[startIdx]` should be non-null. Why would we need to check for anything else?

Reply:
Consider the case, when the user updates the _items_per_page_ . If we check for `publicData[startIdx]` to be non-null then then rows greater than previous _items_per_page_ will be empty.

Reply:
That's a very interesting point; but if that happens isn't it also possible that the "top" and "bottom" of the new "page" are non-blank (the end of one sparse page under the old page size and the beginning of another) but there's a gap in the middle that hadn't been fetched? How will the UI handle that? This code won't fetch that missing data. Maybe this is just theoretical; I know you're only supporting a fixed set of page sizes. But, logically, it seems that, to be safe, you need to fetch the page if *any* entry in the range is empty rather than only if *all* of them are empty.

Reply:
Updated